File: src/Zametek.Maths.Graphs.Primitives/Entities/Graph.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Zametek.Maths.Graphs
{
    /// <summary>
    /// A raw directed-graph structure: the edges and nodes with their content payloads, as exported by the graph builders. Equality compares the full edge and node sets.
    /// </summary>
    public class Graph<T, TEdgeContent, TNodeContent>
        : IEquatable<Graph<T, TEdgeContent, TNodeContent>>
        where T : struct, IComparable<T>, IEquatable<T>
        where TEdgeContent : IHaveId<T>, ICloneObject
        where TNodeContent : IHaveId<T>, ICloneObject
    {
        #region Fields

        private const int HashFactorOne = 17;
        private const int HashFactorTwo = 23;

        #endregion

        #region Ctors

        /// <summary>
        /// Creates an empty graph.
        /// </summary>
        public Graph()
            : this(Enumerable.Empty<Edge<T, TEdgeContent>>(), Enumerable.Empty<Node<T, TNodeContent>>())
        {
        }

        /// <summary>
        /// Creates a graph from the given edges and nodes.
        /// </summary>
        public Graph(IEnumerable<Edge<T, TEdgeContent>> edges, IEnumerable<Node<T, TNodeContent>> nodes)
        {
            if (edges is null)
            {
                throw new ArgumentNullException(nameof(edges));
            }
            if (nodes is null)
            {
                throw new ArgumentNullException(nameof(nodes));
            }
            Edges = edges.ToList();
            Nodes = nodes.ToList();
        }

        #endregion

        #region Properties

        /// <summary>
        /// The edges of the graph.
        /// </summary>
        public IList<Edge<T, TEdgeContent>> Edges
        {
            get;
        }

        /// <summary>
        /// The nodes of the graph.
        /// </summary>
        public IList<Node<T, TNodeContent>> Nodes
        {
            get;
        }

        #endregion

        #region Overrides

        /// <inheritdoc/>
        public override bool Equals(object? obj)
        {
            return Equals(obj as Graph<T, TEdgeContent, TNodeContent>);
        }

        /// <inheritdoc/>
        public override int GetHashCode()
        {
            unchecked
            {
                int hash = HashFactorOne;
                hash = Edges.Select(x => x.Id).OrderBy(x => x).Aggregate(hash, (a, b) => a * HashFactorTwo + b.GetHashCode());
                hash = Nodes.Select(x => x.Id).OrderBy(x => x).Aggregate(hash, (a, b) => a * HashFactorTwo + b.GetHashCode());
                return hash;
            }
        }

        #endregion

        #region IEquatable

        /// <inheritdoc/>
        public bool Equals(Graph<T, TEdgeContent, TNodeContent>? other)
        {
            if (other is null)
            {
                return false;
            }
            return Edges.OrderBy(x => x.Id).SequenceEqual(other.Edges.OrderBy(x => x.Id))
                && Nodes.OrderBy(x => x.Id).SequenceEqual(other.Nodes.OrderBy(x => x.Id));
        }

        #endregion
    }
}

File: rust/primitives/src/graph.rs
use crate::edge::{Edge, HasId};
use crate::key::Key;
use crate::node::Node;

/// A raw directed-graph structure: the edges and nodes with their content
/// payloads, as exported by the graph builders. Equality compares the full
/// edge and node sets (order-insensitive), as in C#.
#[derive(Debug, Clone)]
pub struct Graph<K: Key, EC: HasId<K>, NC: HasId<K>> {
    pub edges: Vec<Edge<K, EC>>,
    pub nodes: Vec<Node<K, NC>>,
}

impl<K: Key, EC: HasId<K>, NC: HasId<K>> Graph<K, EC, NC> {
    /// Creates an empty graph.
    pub fn new() -> Self {
        Self {
            edges: Vec::new(),
            nodes: Vec::new(),
        }
    }

    /// Creates a graph from the given edges and nodes.
    pub fn with_content(edges: Vec<Edge<K, EC>>, nodes: Vec<Node<K, NC>>) -> Self {
        Self { edges, nodes }
    }
}

impl<K: Key, EC: HasId<K>, NC: HasId<K>> Default for Graph<K, EC, NC> {
    fn default() -> Self {
        Self::new()
    }
}

impl<K: Key, EC: HasId<K>, NC: HasId<K>> PartialEq for Graph<K, EC, NC> {
    fn eq(&self, other: &Self) -> bool {
        if self.edges.len() != other.edges.len() || self.nodes.len() != other.nodes.len() {
            return false;
        }
        let sorted_ids = |edges: &[Edge<K, EC>]| {
            let mut ids: Vec<K> = edges.iter().map(|e| e.id()).collect();
            ids.sort();
            ids
        };
        if sorted_ids(&self.edges) != sorted_ids(&other.edges) {
            return false;
        }
        let mut a: Vec<&Node<K, NC>> = self.nodes.iter().collect();
        let mut b: Vec<&Node<K, NC>> = other.nodes.iter().collect();
        a.sort_by_key(|n| n.id());
        b.sort_by_key(|n| n.id());
        a.iter().zip(b.iter()).all(|(x, y)| x == y)
    }
}

impl<K: Key, EC: HasId<K>, NC: HasId<K>> Eq for Graph<K, EC, NC> {}
